public/app/plugins/datasource/prometheus/datasource.ts

```diff
@@ -141,34 +145,122 @@ export class PrometheusDatasource extends DataSourceApi<PromQuery, PromOptions>
+    return series;
+  };
 
-    options = _.clone(options);
+  runObserverQueries = (
+    options: DataQueryRequest<PromQuery>,
+    observer: DataStreamObserver,
+    queries: PromQueryRequest[],
+    activeTargets: PromQuery[],
+    end: number
+  ) => {
+    queries.forEach((query, index) => {
+      const target = activeTargets[index];
+      let observable: Observable<any> = null;
+
+      if (query.instant) {
+        observable = from(this.performInstantQuery(query, end));
+      } else {
+        observable = from(this.performTimeSeriesQuery(query, query.start, query.end));
+      }
+
+      observable
+        .pipe(
+          single(), // unsubscribes automatically after first result
+          filter((response: any) => (response.cancelled ? false : true)),
+          map((response: any) => {
+            return this.processResult(response, query, target, queries.length);
+          })
+        )
+        .subscribe({
+          next: series => {
+            observer({
+              key: `prometheus-${target.refId}`,
+              state: LoadingState.Done,
+              request: options,
+              series,
+              unsubscribe: () => undefined,
+            });
+          },
+        });
+    });
+  };
+
+  prepareTargets = (options: DataQueryRequest<PromQuery>, start: number, end: number) => {
+    const queries: PromQueryRequest[] = [];
+    const activeTargets: PromQuery[] = [];
 
     for (const target of options.targets) {
       if (!target.expr || target.hide) {
         continue;
       }
 
+      if (target.container === 'explore') {
+        target.format = 'time_series';
+        target.instant = false;
+        const instantTarget: any = _.cloneDeep(target);
+        instantTarget.format = 'table';
+        instantTarget.instant = true;
+        instantTarget.valueWithRefId = true;
+        delete instantTarget.maxDataPoints;
+        instantTarget.requestId += '_instant';
+        instantTarget.refId += '_instant';
+        activeTargets.push(instantTarget);
+        queries.push(this.createQuery(instantTarget, options, start, end));
+      }
+
       activeTargets.push(target);
       queries.push(this.createQuery(target, options, start, end));
     }
 
+    return {
+      queries,
+      activeTargets,
+    };
+  };
+
+  query(options: DataQueryRequest<PromQuery>, observer?: DataStreamObserver): Promise<{ data: any }> {
+    const start = this.getPrometheusTime(options.range.from, false);
+    const end = this.getPrometheusTime(options.range.to, true);
+
+    options = _.clone(options);
+    const { queries, activeTargets } = this.prepareTargets(options, start, end);
+
     // No valid targets, return the empty result to save a round trip.
     if (_.isEmpty(queries)) {
       return this.$q.when({ data: [] }) as Promise<{ data: any }>;
     }
 
+    if (
+      observer &&
+      options.targets.filter(target => target.container === 'explore').length === options.targets.length
```

Comment:
This if statement is so that we can return table data instantly but I couldn't get Prometheus Panel to show graph with this so only using this in explore. Would like to know more specifically what the observer in PanelRunner expects.

Reply:
@hugohaggmark -- sorry I missed this comment/question.  Its great to have new people using this and hopefully we can improve the api/docs/comments as we solve confusion :)

Without knowing what your issue was, I'm guessing some issue with the`key` or `request` fields in the state response:
https://github.com/grafana/grafana/blob/master/packages/grafana-ui/src/types/datasource.ts#L272

The request is used to know if the update came from a query we asked for or is from an old request.  see [shouldDisconnect()](https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryState.ts#L300)

Also note that the `series` should contain everything with the same key.  If you have one query that returns two series in two requests, you should do something like:

1.  quickly return: `{series: [first]}`
2. run the slower query, and return both: `{series: [first,second]}`

Each response with the same key will replace the entire series[], so that state should be managed by the DataSource